### PR TITLE
Faster promoting arithmetic

### PR DESCRIPTION
**Pull request by timholy**
Reference: #21065. The main reason to introduce this is so that it can be backported to 0.5; perhaps on 0.6 there can be a deeper solution. I could re-file this explicitly against `release-0.5` if that's preferable.

### CODE REVIEW

**Comment by martinholters** [comment]
How about this instead:
```diff
--- a/base/promotion.jl
+++ b/base/promotion.jl
@@ -171,6 +171,7 @@ promote_result{T,S}(::Type{T},::Type{S},::Type{Bottom},::Type{Bottom}) = (@_pure
 promote() = ()
 promote(x) = (x,)
 function promote{T,S}(x::T, y::S)
+    @_inline_meta
     (convert(promote_type(T,S),x), convert(promote_type(T,S),y))
 end
 promote_typeof(x) = (@_pure_meta; typeof(x))
```

**Comment by timholy** [comment]
I wondered about that. There are occasionally weird things that can happen performance-wise when you change the "level" that gets inlined (e.g., https://github.com/JuliaImages/ImageCore.jl/pull/26). But I guess with nanosoldier there is little reason not to try and find out what happens.

**Comment by martinholters** [comment]
Let's ask @nanosoldier `runbenchmarks(ALL, vs = ":master")` and then again for the inlining of `promote` version and see whether that gives any insights how better to proceed.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21067#issuecomment-287366816) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/1de7d0c2ec49c5dd25cc53f43a262d1b6ca4feaa/545555f_vs_b549d2b/report.md). cc @jrevels

**Comment by stevengj** [comment]
It seems like it would be better to fix `promote(x,y)...` inlining.

**Comment by KristofferC** [comment]
IIUC, that is being fixed for 0.6 and this is a minimal PR towards 0.5.2?

**Comment by stevengj** [comment]
This shouldn't this be a PR against the 0.5 branch rather than master?

**Comment by timholy** [comment]
> IIUC, that is being fixed for 0.6 and this is a minimal PR towards 0.5.2?

Sort of. Forced inlining of `promote` is not currently part of #21069. I'll submit a version that does both just so we can get benchmarks, and then generate separate PRs against master and release-0.5.

**Comment by timholy** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

File: base/promotion.jl [hunk 1]
```diff
@@ -243,10 +245,22 @@ function sametype_error(input...)
           " failed to change any input types")
 end
 
-+(x::Number, y::Number) = +(promote(x,y)...)
-*(x::Number, y::Number) = *(promote(x,y)...)
--(x::Number, y::Number) = -(promote(x,y)...)
-/(x::Number, y::Number) = /(promote(x,y)...)
+function +(x::Number, y::Number)
+    xp, yp = promote(x, y)
```
Comment (martinholters) [inline base/promotion.jl]: Are these changes necessary in addition to the inlining of `promote`?

Comment (StefanKarpinski) [inline base/promotion.jl]: Yes, I really don't like having to expand these like this – it's very ugly and we've never had to do so before.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21067#issuecomment-287428073) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/2ffb074ee8f75860c990503d6c09fa96f314bda5/6d01127_vs_762b943/report.md). cc @jrevels

**Comment by martinholters** [comment]
So maybe inlining `promote` is not such a good idea after all...

**Comment by timholy** [comment]
Now that's what I'd call an uninspiring benchmark result. I expected some regressions, but that really took the idea and ran with it.

File: base/promotion.jl (hunk 1 above)
Comment (timholy) [inline base/promotion.jl]: Certainly much less important with #21069. The main problem is that as we implement more abstractions, the inlining heuristics sometimes seem to fail to realize how simple something really is. I don't think we can exclude the possibility that we're noticing problems that we didn't notice before.

**Comment by martinholters** [comment]
So back to just rewriting `+`, `*`, `-`, `/` as done in the first commit it is, I guess. Is there enough benefit to do that in addition to #21069? Or should we do that just for 0.5? And/or should we backport #21069 to 0.5?

**Comment by timholy** [comment]
I'd be content with #21069 alone, if it is backportable. It's certainly the most far-reaching of the solutions. 

Adding the first commit is another 35% faster on this particular benchmark; let's see what others think of whether that's worth it.

**Comment by martinholters** [comment]
Backport of #21069 at and #21080.
